### ISSUE TITLE
Intellisense retention

### ISSUE THREAD

**Issue by SteelBlueVision** [microsoft/vscode-cpptools#7372]
For a directory workspace based project:

Is it possible to force Intellisense to run on all files in a particular folder, without having to open said files and to persist this information from run to run of VS Code so that it does not have to be constantly regenerated when files get closed and reopened or when VS Code itself is closed and reopened.

I've maxed out all Intellisense config settings in an effort to retain as much info as possible:

  "C_Cpp.intelliSenseCacheSize": 32768,
  "C_Cpp.intelliSenseMemoryLimit": 16384,

I don't want any background parsing, tagging, sensing, or whatever that seems to occur again and again, each time I open the same file (even in the same VS Code session, after it has already been parsed countless times and has not changed between parses). I want to be able to do a parse once, wait for it to complete, and have it recur **only** on file change, thereafter. Whatever intellisense info gets generated, I want it persisted to the ipch area and available the next time it is needed. I want to be able to open source files that have already had Intellisense generated for them, to open fully "Intellisensed" when I open them and not have to wait each and every single time (sometimes for several seconds) for Intellisense to parse them and regenerate some info when I open them, that it already generated the last time I opened them (and the files have not changed).

In other words, rather than activating Intellisense on files when they are opened (i.e., on demand), I want to force it to run on certain directories on demand and retain this info permanently (unless the files change, of course).

**Comment by sean-mcmanus** [comment]
Unfortunately, this is not possible with our design/implementation. The best you can do is to use a machine with 64 cores (and enough RAM) and keep up to 64 files open without reloading your workspace.

The IntelliSense cache (ipch) can help somewhat with reducing parse time for includes.

You could try filing a feature request at https://developercommunity.visualstudio.com/report?space=62&entry=suggestion on VS, but they might close it as "by design" or "won't fix".

**Comment by sean-mcmanus** [comment]
Actually, I think we may start to shutdown TUs after they start using > 1 GB RAM in total -- let us know if you're hitting that and we may be able to add a fix/workaround.

**Comment by mgoldshteyn** [comment]
When you say TUs using > 1 GB RAM, how do I determine this? Is it based on the size of the actual source files or based on the size of the in-memory Intellisense info? Or where in VS Code and/or Linux can I look to see how much memory is being used "for open TUs?"

**Comment by sean-mcmanus** [comment]
I'm referring to the memory used by cpptools-srv. You can use the "top" command in Linux. It should be the RES column.

**Comment by sean-mcmanus** [comment]
The C_Cpp.intelliSenseMemoryLimit is for a single TU, but we don't have a setting yet for the combined TU's.
